File: packages/domain/src/adapters/drift.ts
/**
 * Parser drift metrics: types and recording helpers.
 */
import type { EvaluateRunHealthInput, EvaluateRunHealthResult } from './run-health.js';

export type ParserDriftMetric = {
  readonly adapterId: string;
  readonly parserVersion: string;
  readonly registryEntryId: string;
  readonly runId: string;
  readonly recordedAt: string;
  readonly expectedRecordCount: number;
  readonly actualRecordCount: number;
  readonly expectedSchemaVersion: string;
  readonly observedSchemaVersion: string;
  readonly fieldNullRates: Readonly<Record<string, number>>;
  readonly issues: readonly string[];
};

export type DriftAccumulator = {
  readonly adapterId: string;
  readonly parserVersion: string;
  readonly registryEntryId: string;
  readonly runId: string;
  readonly startedAt: string;
  fieldNullCounts: Record<string, { nulls: number; total: number }>;
};

export function createDriftAccumulator(input: {
  readonly adapterId: string;
  readonly parserVersion: string;
  readonly registryEntryId: string;
  readonly runId: string;
  readonly startedAt: string;
}): DriftAccumulator {
  return {
    ...input,
    fieldNullCounts: {},
  };
}

export function recordFieldObservation(
  accumulator: DriftAccumulator,
  field: string,
  isNull: boolean,
): void {
  const current = accumulator.fieldNullCounts[field] ?? { nulls: 0, total: 0 };
  accumulator.fieldNullCounts[field] = {
    nulls: current.nulls + (isNull ? 1 : 0),
    total: current.total + 1,
  };
}

export function computeFieldNullRates(
  accumulator: DriftAccumulator,
): Readonly<Record<string, number>> {
  const rates: Record<string, number> = {};
  for (const [field, counts] of Object.entries(accumulator.fieldNullCounts)) {
    rates[field] = counts.total === 0 ? 0 : counts.nulls / counts.total;
  }
  return rates;
}

export function buildParserDriftMetric(
  accumulator: DriftAccumulator,
  input: EvaluateRunHealthInput,
  health: EvaluateRunHealthResult,
  recordedAt: string,
): ParserDriftMetric {
  return {
    adapterId: accumulator.adapterId,
    parserVersion: accumulator.parserVersion,
    registryEntryId: accumulator.registryEntryId,
    runId: accumulator.runId,
    recordedAt,
    expectedRecordCount: input.expectedCount,
    actualRecordCount: input.actualCount,
    expectedSchemaVersion: input.expectedSchemaVersion,
    observedSchemaVersion: input.observedSchemaVersion,
    fieldNullRates: computeFieldNullRates(accumulator),
    issues: health.details,
  };
}

export type DriftMetricStore = {
  list(adapterId?: string): readonly ParserDriftMetric[];
  record(metric: ParserDriftMetric): void;
};

export function createInMemoryDriftMetricStore(
  seed: readonly ParserDriftMetric[] = [],
): DriftMetricStore {
  const metrics: ParserDriftMetric[] = [...seed];
  return {
    list(adapterId?: string) {
      if (!adapterId) {
        return metrics;
      }
      return metrics.filter((metric) => metric.adapterId === adapterId);
    },
    record(metric: ParserDriftMetric) {
      metrics.push(metric);
    },
  };
}

File: workers/research/src/black_book_research/adapters/drift.py
"""Parser drift metrics mirroring domain drift helpers (BB-037)."""

from __future__ import annotations

from dataclasses import dataclass, field

from .types import DriftAccumulator, EvaluateRunHealthResult, ParserDriftMetric


@dataclass(slots=True)
class InMemoryDriftMetricStore:
    metrics: list[ParserDriftMetric] = field(default_factory=list)

    def list(self, adapter_id: str | None = None) -> list[ParserDriftMetric]:
        if adapter_id is None:
            return list(self.metrics)
        return [metric for metric in self.metrics if metric.adapter_id == adapter_id]

    def record(self, metric: ParserDriftMetric) -> None:
        self.metrics.append(metric)


def create_drift_accumulator(
    *,
    adapter_id: str,
    parser_version: str,
    registry_entry_id: str,
    run_id: str,
    started_at: str,
) -> DriftAccumulator:
    return DriftAccumulator(
        adapter_id=adapter_id,
        parser_version=parser_version,
        registry_entry_id=registry_entry_id,
        run_id=run_id,
        started_at=started_at,
    )


def record_field_observation(
    accumulator: DriftAccumulator, field_name: str, is_null: bool
) -> None:
    counts = accumulator.field_null_counts.setdefault(field_name, {"nulls": 0, "total": 0})
    counts["nulls"] += 1 if is_null else 0
    counts["total"] += 1


def compute_field_null_rates(accumulator: DriftAccumulator) -> dict[str, float]:
    rates: dict[str, float] = {}
    for field_name, counts in accumulator.field_null_counts.items():
        total = counts["total"]
        rates[field_name] = 0.0 if total == 0 else counts["nulls"] / total
    return rates


def build_parser_drift_metric(
    accumulator: DriftAccumulator,
    *,
    expected_count: int,
    actual_count: int,
    expected_schema_version: str,
    observed_schema_version: str,
    health: EvaluateRunHealthResult,
    recorded_at: str,
) -> ParserDriftMetric:
    return ParserDriftMetric(
        adapter_id=accumulator.adapter_id,
        parser_version=accumulator.parser_version,
        registry_entry_id=accumulator.registry_entry_id,
        run_id=accumulator.run_id,
        recorded_at=recorded_at,
        expected_record_count=expected_count,
        actual_record_count=actual_count,
        expected_schema_version=expected_schema_version,
        observed_schema_version=observed_schema_version,
        field_null_rates=compute_field_null_rates(accumulator),
        issues=health.details,
    )
